src/python/pants/backend/python/lint/pyupgrade/rules.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pants.backend.python.lint.python_fmt import PythonFmtRequest
+from pants.backend.python.lint.pyupgrade.skip_field import SkipPyUpgradeField
+from pants.backend.python.lint.pyupgrade.subsystem import PyUpgrade
+from pants.backend.python.target_types import PythonSources
+from pants.backend.python.util_rules import pex
+from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, MergeDigests
+from pants.engine.process import FallibleProcessResult, Process
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class PyUpgradeFieldSet(FieldSet):
+    required_fields = (PythonSources,)
+
+    sources: PythonSources
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipPyUpgradeField).value
+
+
+class PyUpgradeRequest(PythonFmtRequest, LintRequest):
+    field_set_type = PyUpgradeFieldSet
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: PyUpgradeRequest
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+def generate_argv(source_files: SourceFiles, pyupgrade: PyUpgrade) -> tuple[str, ...]:
+    args = [*pyupgrade.args]
+    args.extend(source_files.files)
+    return tuple(args)
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_pyupgrade(setup_request: SetupRequest, pyupgrade: PyUpgrade) -> Setup:
+    request = setup_request.request
+    pyupgrade_pex_get = Get(
+        VenvPex,
+        PexRequest(
+            output_filename="pyupgrade.pex",
+            internal_only=True,
+            requirements=pyupgrade.pex_requirements(),
+            interpreter_constraints=pyupgrade.interpreter_constraints,
```

Comment:
I think that this is subject to the same issue as https://github.com/pantsbuild/pants/pull/13010#issuecomment-930761734: this tool will need to be built with the interpreter constraints of the code that it is operating on.

I expect that that can be resolved, but let's hold off on landing this until we have a clear path forward there.

Reply:
Unlike other tools, the interpreter running py upgrade is irrelevant to what kind of changes it makes or checks for. This part is controlled by the command line flags passed to it.
So as long as the interpreter is 3.6.1 or higher it should work.
I chose to specify 3.7 or higher because 3.6 is deprecated

Reply:
Are you certain that running with Py36 can still parse Py38+ syntax? Most tools cannot do that, like Flake8. But some like Black and MyPy can.

Reply:
I didn't check.
But I do know that it can read python 2.x code and upgrade it to 3.x so I would assume the python interpreter is less relevant with this tool, I am not even sure it uses ast, I think it uses simple string tokenization to parse stuff.
but I could be wrong.